lib/puppet_pal.rb

```diff
@@ -201,6 +205,15 @@ def self.prepare_node_facts(node, facts)
     end
   end
 
+  def self.add_variables(scope, variables)
+    return if variables.nil?
+    unless variables.is_a?(Hash)
+      raise Argument_Error, _("Given variables must be a hash, got %{type}") % { type: variables.class }
```

Comment:
I would expect a check for `Hash[Pattern[<VAR_NAME>],RichData]` here to avoid that `Runtime` values and illegal variable names in the evaluation (`<VAR_NAME>` would be picked from _lib/puppet/pops/patterns.rb_).